tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,949 @@
+//Copyright(c) Microsoft.All rights reserved.
+//Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifdef __cplusplus
+#include <cstdlib>
+#include <cstddef>
+#include <cstdint>
+#else
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdint.h>
+#endif
+
+
+#include "testrunnerswitcher.h"
+
+static TEST_MUTEX_HANDLE g_testByTest;
+
+
+
+#include "azure_c_util/interlocked.h"
+
+
+BEGIN_TEST_SUITE(interlocked_int)
+
+
+TEST_SUITE_INITIALIZE(a)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+}
+
+TEST_SUITE_CLEANUP(b)
+{
+
+    TEST_MUTEX_DESTROY(g_testByTest);
+}
+
+TEST_FUNCTION_INITIALIZE(c)
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+}
+
+TEST_FUNCTION_CLEANUP(d)
+{
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/*Tests_SRS_INTERLOCKED_43_001: [ interlocked_add shall atomically add *addend with value and store the result in *addend.]*/
+/*Tests_SRS_INTERLOCKED_43_032: [interlocked_add shall return the result of the addition.]*/
+TEST_FUNCTION(interlocked_add_does_addition)
+{
+    ///arrange
+    volatile int32_t addend = INT32_MAX;
+    int32_t value = INT32_MIN;
+
+    ///act
+    int32_t return_val = interlocked_add(&addend, value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, -1, interlocked_or((volatile int32_t*)&addend, 0));
```

Comment:
>(volatile int32_t*) [](start = 49, length = 19)

Unnecessary cast.

Also, interlocked_add is more of a convention for reading (obviously not available for 16/8-bit, but should use interlocked_add/interlocked_add_64 where possible). /consider

Reply:
I believe you can remove the casts here and everywhere in this file

---
In reply to: [428944893](https://github.com/Azure/azure-c-util/pull/43#discussion_r428944893) [](ancestors = 428944893)

Reply:
I think it would be better to do all the reads with interlocked_or so that things are consistent through the file. 

---
In reply to: [428944893](https://github.com/Azure/azure-c-util/pull/43#discussion_r428944893) [](ancestors = 428944893)

Reply:
This cast can be removed, but in the places where a pointer to an unsigned integer is being passed, a cast must be used.

---
In reply to: [428945102](https://github.com/Azure/azure-c-util/pull/43#discussion_r428945102) [](ancestors = 428945102,428944893)